add-interactive.c

```diff
@@ -0,0 +1,188 @@
+	struct adddel index, worktree;
+	char name[FLEX_ARRAY];
+};
+
+struct collection_status {
+	int collecting_from_index;
+
+	const char *reference;
+	struct pathspec pathspec;
+
+	struct hashmap file_map;
+};
+
+static int pathname_equal(const void *unused_cmp_data,
+			  const void *entry, const void *entry_or_key,
+			  const void *keydata)
+{
+	const struct file_stat *e1 = entry, *e2 = entry_or_key;
+	const char *name = keydata ? keydata : e2->name;
+
+	return strcmp(e1->name, name);
+}
+
+static int pathname_cmp(const void *a, const void *b)
+{
+	struct file_stat *f1 = *((struct file_stat **)a);
+	struct file_stat *f2 = *((struct file_stat **)b);
+
+	return strcmp(f1->name, f2->name);
+}
+
+static void populate_adddel(struct adddel *ad, uintmax_t add, uintmax_t del)
+{
+	ad->add = add;
+	ad->del = del;
+}
+
+static void collect_changes_cb(struct diff_queue_struct *q,
+			       struct diff_options *options,
+			       void *data)
+{
+	struct collection_status *s = data;
+	struct diffstat_t stat = { 0 };
+	int i;
+
+	if (!q->nr)
+		return;
+
+	compute_diffstat(options, &stat, q);
+
+	for (i = 0; i < stat.nr; i++) {
+		struct file_stat *entry;
+		const char *name = stat.files[i]->name;
+		unsigned int hash = strhash(name);
+
+		entry = hashmap_get_from_hash(&s->file_map, hash, name);
+		if (!entry) {
+			FLEX_ALLOC_STR(entry, name, name);
+			hashmap_entry_init(entry, hash);
+			hashmap_add(&s->file_map, entry);
+		}
+
+		if (s->collecting_from_index)
+			populate_adddel(&entry->index, stat.files[i]->added,stat.files[i]->deleted);
+		else
+			populate_adddel(&entry->worktree, stat.files[i]->added, stat.files[i]->deleted);
+	}
+}
+
+static void collect_changes_worktree(struct collection_status *s)
+{
+	struct rev_info rev;
+
+	s->collecting_from_index = 0;
+
+	init_revisions(&rev, NULL);
+	setup_revisions(0, NULL, &rev, NULL);
+
+	/* Use the max_count field to specify the unmerged stage, against
+	 * which the working tree file is compared for an unmerged path.
+	 * git diff-files itself when running cmd_diff_files() leaves
+	 * rev.max_count untouched to get a normal output (as opposed to
+	 * the case when it is told to do --base/--ours/--theirs), so it
+	 * ends up passing -1 in this field in such a case.
+	 */
+	rev.max_count = 0;
+
+	rev.diffopt.flags.ignore_dirty_submodules = 1;
+	rev.diffopt.output_format = DIFF_FORMAT_CALLBACK;
+	rev.diffopt.format_callback = collect_changes_cb;
+	rev.diffopt.format_callback_data = s;
+
+	run_diff_files(&rev, 0);
+}
+
+static void collect_changes_index(struct collection_status *s)
+{
+	struct rev_info rev;
+	struct setup_revision_opt opt = { 0 };
+
+	s->collecting_from_index = 1;
+
+	init_revisions(&rev, NULL);
+	opt.def = s->reference;
+	setup_revisions(0, NULL, &rev, &opt);
+
+	rev.diffopt.output_format = DIFF_FORMAT_CALLBACK;
+	rev.diffopt.format_callback = collect_changes_cb;
+	rev.diffopt.format_callback_data = s;
+
+	run_diff_index(&rev, 1);
+}
+
+static int on_unborn_branch(void)

```

Comment:
I had a look at this documentation: https://github.com/git/git/blob/v2.21.0/refs.h#L12-L70

From that, it would appear to me that we could pass in `&oid` as 3rd parameter and avoid the `get_oid()` call altogether.

Also, passing in a `&flags` as fourth parameter would allow us to see whether it was a symbolic ref by testing that `flags` for the `REF_ISSYMREF` bit after `resolve_ref_unsafe()` succeeded.

And I think the second parameter should be `RESOLVE_REF_READING`.